Initial support for PDF/A-1a/b

andreiaugustin opened the pull request:
<!--
Please make sure you are familiar with and follow the instructions in the
contributing guidelines (found in the CONTRIBUTING.md file).

Please fill out the information below to expedite the review and (hopefully)
merge of your pull request!
-->

<!-- Is it a Bug fix, feature, docs update, ... -->
Feature
<!-- You can also link to an open issue here -->
**What kind of change does this PR introduce?**

These changes add initial support for PDF/A-1a and PDF/A-1b which targets a profile for long-term archiving.
In order to create PDF/A files, set 'pdfa' to either '1a' or '1b' in the options of the PDFDocument.

Note: PDF/A does not support encryption and other things detailed in the docs or in the standard. This change does not prevent you from creating encrypted PDFs even if pdfa option is enabled - in those cases, your PDF may not pass PDF/A validation. This means it's down to the user to ensure all other features that are restricted from PDF/A standard are not enabled.

<!-- Have you done all of these things?  -->
**Checklist**:

<!-- add "N/A" to the end of each line that's irrelevant to your changes -->
<!-- to check an item, place an "x" in the box like so: "- [x] Documentation" -->

- [x] Unit Tests
- [x] Documentation
- [x] Update CHANGELOG.md
- [x] Ready to be merged <!-- In your opinion, is this ready to be merged as soon as it's reviewed? -->

<!-- feel free to add additional comments -->

This change will currently support only PDF/A 1b and 1a ( 1a only when tagged is set to true ). If you're happy with the changes, I can continue adding support for other PDF/A standards: 2a, 2b, 2u, 3a, 3b, 3u, 4e, 4f.

devongovett commented on lib/mixins/pdfa.js:
```diff
@@ -0,0 +1,110 @@
+export default {
+
+    initPDFA(pOptions) {
+        if (pOptions.pdfa) {
```
Maybe this option should be called `profile`, in case we want to add support for other PDF profiles like `PDF/E`, `PDF/UA`, etc. in the future? Could accept a string like `profile: 'PDF/A-1'` etc.

andreiaugustin replied:
Great idea - I can see us needing support for PDF/UA very soon and would be nice to support PDF/E and other subsets in the future.

Would you be happy with `profile` or `subset`/`pdfsubset`?

Also, would you prefer receiving strings or shall we have identifiers and therefore pass something along the lines of `PDFDocument.Subsets.PDFA1a`/`PDFDocument.Subsets.PDFA1b`/`PDFDocument.Subsets.PDFE`? I'm thinking they might be cleaner than strings, plus if we ever come across a case where a PDF can be PDF/A and PDF/E compliant, we can pass `PDFDocument.Subsets.PDFA1a+PDFDocument.Subsets.PDFE`?

devongovett replied:
Personally, I prefer strings for brevity, but we could also support an array if multiple can be supported at once. I don't have a strong preference on the name. Subsets are fine.

andreiaugustin replied:
Added a new commit to support `subset` as well as a couple mixins to control metadata and load subsets at runtime, if you're happy I can continue and add the documentation for this?

andreiaugustin replied:
Docs should now refer to `subset` instead of the old `pdfa` :)

Is there anything else I should check/change or is this good to merge?

andreiaugustin commented:
@blikblum 

Thanks! I agree - XML metadata should be able to replace the /Info, but I think metadata streams were added only in 1.4, therefore we'd still need to use /Info for 1.3?

Yes we can certainly store the colour profile separately - will create a folder and we can have each in their own file.

In terms of testing, I think the most reliable way would be to run verapdf's binary tool to check against the subset/profile defined in the generated pdf - otherwise we quickly find ourselves re-implementing every test and reinventing the wheel (e.g. PDF/A-1: https://github.com/veraPDF/veraPDF-validation-profiles/wiki/PDFA-Part-1-rules/)

andreiaugustin commented:
@blikblum 

I've added a more general mixin for controlling the metadata and changed it so pdf versions 1.3 uses /Info and pdf versions >1.3 do not show a /Info but rather store the info as metadata.

Although it works fine and passes PDF/A-1 validation, software like Finder's 'Get info' does not seem to parse the info stored as metadata on Monterey 12.6. 

Even if it's down to each implementation to decide whether to follow the standard, I think the extra space to store the info both in /Info and /Metadata might be worth it in order to support implementation that do not parse the metadata?

blikblum commented:
AFAIR from reading the spec is fine having both /Info and /Metadata

> In terms of testing

The testing is just checking is the produced output is the intended (see other tests), mostly to avoid regressions in future development

andreiaugustin commented:
@blikblum

Should be done now! Added tests for metadata and pdfa1 subset, plus we now have /Info and /Metadata for pdf version >=1.4 and only /Info for pdf version 1.3 (added test for this too in document.spec.js).

Trim commented on lib/color_profiles/sRGB_IEC61966_ICC.js:
```diff
@@ -0,0 +1,4 @@
+export default Buffer(
+    'AAAL7AAAAAACAAAAbW50clJHQiBYWVogB9kAAwAbABUAJQAtYWNzcAAAAAAAAAAAAAAAAAAAAAAAAAABAAAAAAAAAAEAAPbWAAEAAAAA0y0AAAAAyVvWN+ldijsN84+ZwTIDiQAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAQZGVzYwAAAUQAAAB9YlhZWgAAAcQAAAAUYlRSQwAAAdgAAAgMZG1kZAAACeQAAACIZ1hZWgAACmwAAAAUZ1RSQwAAAdgAAAgMbHVtaQAACoAAAAAUbWVhcwAACpQAAAAkYmtwdAAACrgAAAAUclhZWgAACswAAAAUclRSQwAAAdgAAAgMdGVjaAAACuAAAAAMdnVlZAAACuwAAACHd3RwdAAAC3 ...
```
I'm just a user of pdfkit not a developper, but I find a bit strange to write a base64 encoded string to create a buffer.

Is it possible to store the color profile in a binary file and use the [fs.readFileSync('./sRGB_IEC61966.icc', { encoding: null, flags: 'r'})](https://nodejs.org/api/fs.html#fsreadfilesyncpath-options) to create the Buffer directly ?

andreiaugustin replied:
Added a new commit which now stores the sRGB 2.1 colour profile in the data and loads it via readFileSync :)

andreiaugustin commented:
Hey guys, sorry to be pushy - would it be possible to merge this if you're all happy? Was hoping to include the PDF/A support in our next Beta and didn't want to use a fork :/

Trim commented:
I'm not a developer of pdfkit, so my approval for this request will not have any value.

As I'm just a user of pdfkit, I can't help you to make progress, sorry.

KingNoosh commented:
@blikblum Any other feedback for this PR to get approved?

blikblum commented:
Looks pretty good to me. I'll merge and wait a week or so before doing the release to see if other maintainers have objections

The only improvement i see in future is loading the profile info in demand so those that don't need can save a few kb
